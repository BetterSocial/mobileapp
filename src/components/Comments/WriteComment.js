import * as React from 'react';
import FastImage from 'react-native-fast-image';
import ToggleSwitch from 'toggle-switch-react-native';
import {ActivityIndicator, StyleSheet, Text, TextInput, TouchableOpacity, View} from 'react-native';

import AsyncStorage from '@react-native-async-storage/async-storage';
import AnonUserInfoRepo from '../../service/repo/anonUserInfoRepo';
import MemoSendComment from '../../assets/icon/IconSendComment';
import StringConstant from '../../utils/string/StringConstant';
import {Context} from '../../context';
import {colors} from '../../utils/colors';
import {fonts} from '../../utils/fonts';
import {DEFAULT_PROFILE_PIC_PATH} from '../../utils/constants';

const WriteComment = ({
  value = null,
  onPress,
  onChangeText,
  username,
  inReplyCommentView = false,
  showProfileConnector = true,
  loadingPost = false,
  postId = ''
}) => {
  const [profile] = React.useContext(Context).profile;
  const commentInputRef = React.useRef(null);
  const [isAnonimity, setIsAnonimity] = React.useState(false);
  const isCommentEnabled = value.length > 0;
  const [loadingUser, setLoadingUser] = React.useState(false);
  const isDisableSubmit = !isCommentEnabled || loadingPost;
  const [anonimityData, setAnoimityData] = React.useState({});
  const storageKey = 'isAnonymByDefault';
  const getAnonUser = React.useCallback(async () => {
    setLoadingUser(true);
    try {
      if (!isAnonimity) {
        const response = await AnonUserInfoRepo.getCommentAnonUserInfo(postId);
        setLoadingUser(false);
        if (response?.isSuccess) {
          return setAnoimityData(response.data);
        }

        setIsAnonimity(false);
        return setAnoimityData({});
      }
      setLoadingUser(false);
      return setAnoimityData({});
    } catch (e) {
      console.log(e, 'error');
      return setAnoimityData({});
    }
  }, [isAnonimity, postId]);

  const toggleSwitch = () => {
    setIsAnonimity((prevState) => !prevState);
    getAnonUser();
    if (isAnonimity) saveToStorage('null');
    if (!isAnonimity) saveToStorage('true');
  };

  React.useEffect(() => {
    AsyncStorage.getItem(storageKey).then((data) => {
      if (data === 'true') {
        setIsAnonimity(true);
        getAnonUser();
      }
    });
  }, []);

  const saveToStorage = (valueData) => {
    AsyncStorage.setItem(storageKey, valueData);
  };
  return (
    <View style={styles.columnContainer}>
      <View style={styles.connectorTop(inReplyCommentView, showProfileConnector)} />
      <View style={{flexDirection: 'row', paddingRight: 10}}>
        <Text style={styles.replyToContainer(inReplyCommentView)}>
          <Text style={styles.replyToTitle}>Reply to </Text>
          {username}
        </Text>
        <View style={styles.anonimityContainer}>
          <ToggleSwitch
            isOn={isAnonimity}
            onColor={colors.bondi_blue}
            label="Anonymity"
            offColor="#F5F5F5"
            size="small"
            labelStyle={styles.switch}
            onToggle={toggleSwitch}
          />
        </View>
      </View>
      <View style={styles.container(inReplyCommentView)}>
        <View style={styles.connectorBottom(inReplyCommentView, showProfileConnector)} />
        {isAnonimity ? (
          <>
            {loadingUser ? (
              <ActivityIndicator size={'small'} color={colors.bondi_blue} />
            ) : (
              <View style={[styles.image, {backgroundColor: anonimityData.colorCode}]}>
                <Text style={styles.emojyStyle}>{anonimityData.emojiCode}</Text>
              </View>
            )}
          </>
        ) : (
          <>
            <FastImage
              style={styles.image}
              source={{
                uri: profile.myProfile.profile_pic_path
                  ? profile.myProfile.profile_pic_path
                  : DEFAULT_PROFILE_PIC_PATH
              }}
            />
          </>
        )}

        <TextInput
          testID="changeinput"
          ref={commentInputRef}
          placeholder={StringConstant.commentBoxDefaultPlaceholder}
          placeholderTextColor={colors.gray}
          style={[styles.text, styles.content]}
          onChangeText={onChangeText}
          value={value}
          multiline
          textAlignVertical="center"
        />
        <TouchableOpacity
          testID="iscommentenable"
          onPress={() => onPress(isAnonimity, anonimityData)}
          style={styles.btn(isDisableSubmit || loadingUser)}
          disabled={isDisableSubmit || loadingUser}>
          <MemoSendComment style={styles.icSendButton} />
        </TouchableOpacity>
      </View>
    </View>
  );
};

export default React.memo(WriteComment);

export const styles = StyleSheet.create({
  columnContainer: {
    display: 'flex',
    flexDirection: 'column',
    backgroundColor: colors.white,
    flex: 1,
    width: '100%',
    position: 'absolute',
    bottom: 0,
    borderTopWidth: 1,
    borderTopColor: colors.gray1,
    // zIndex: 1,
    paddingBottom: 14
  },
  replyToContainer: (inReplyCommentView) => ({
    marginLeft: inReplyCommentView ? 90 : 60,
    fontFamily: fonts.inter[600],
    marginBottom: 11,
    marginTop: 7,
    lineHeight: 15,
    fontSize: 12,
    color: colors.gray
  }),
  replyToTitle: {
    fontFamily: fonts.inter[600],
    lineHeight: 15,
    fontSize: 12,
    color: colors.black
  },
  container: (inReplyCommentView) => ({
    flex: 1,
    backgroundColor: '#fff',
    width: '100%',
    paddingRight: 10,
    paddingLeft: inReplyCommentView ? 50 : 20,
    flexDirection: 'row',
    zIndex: 100
  }),
  content: {
    display: 'flex',
    flexDirection: 'column',
    // alignItems: 'center',
    backgroundColor: colors.lightgrey,
    marginLeft: 8,
    borderRadius: 8,
    paddingLeft: 6,
    paddingRight: 8,
    marginEnd: 8,
    flex: 1,
  },
  btn: (isDisableSubmit) => ({
    backgroundColor: !isDisableSubmit ? colors.bondi_blue : '#f2f2f2',
    borderRadius: 18,
    width: 35,
    height: 35,
    display: 'flex',
    justifyContent: 'center',
    marginBottom: 1.5,
    alignSelf: 'flex-end'
  }),
  btnText: {color: 'white', fontFamily: fonts.inter[400]},
  image: {
    width: 36,
    height: 36,
    marginLeft: -7,
    zIndex: -10,
    borderRadius: 18,
    alignItems: 'center',
    justifyContent: 'center'
  },
  text: {
    flex: 1,
    fontSize: 12,
    fontFamily: fonts.inter[400],
    color: colors.black,
    lineHeight: 19,
    paddingTop: 5,
    paddingBottom: 5,
    maxHeight: 100
  },
  icSendButton: {
    alignSelf: 'center'
  },
  connectorTop: (inReplyCommentView, showProfileConnector) => ({
    height: showProfileConnector ? 36 : 0,
    width: 1,
    backgroundColor: colors.gray1,
    position: 'absolute',
    top: 0,
    left: inReplyCommentView ? 60 : 30,
    zIndex: -100
  }),
  connectorBottom: (inReplyCommentView, showProfileConnector) => ({
    height: showProfileConnector ? 20 : 0,
    width: 1,
    backgroundColor: colors.gray1,
    position: 'absolute',
    top: 0,
    left: inReplyCommentView ? 60 : 30,
    zIndex: -100
  }),
  anonimityContainer: {
    marginLeft: 'auto',
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center'
  },
  emojyStyle: {
    fontSize: 18
  },
  switch: {
    fontFamily: fonts.inter[400],
    fontSize: 12,
    color: colors.gray
  }
});
