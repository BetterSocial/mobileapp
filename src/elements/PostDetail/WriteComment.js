import * as React from 'react';
import {
  Image,
  StyleSheet,
  Text,
  TextInput,
  View,
  TouchableOpacity,
} from 'react-native';

import {colors} from '../../utils/colors';
import {fonts} from '../../utils/fonts';

const WriteComment = ({value = null, onPress, onChangeText}) => {
  return (
    <View style={styles.container}>
      <Image
        style={styles.image}
        source={require('../../assets/images/ProfileDefault.png')}
      />
      <View
        style={{
          flexDirection: 'row',
          alignItems: 'center',
          backgroundColor: colors.lightgrey,
          marginLeft: 10,
          borderRadius: 5,
          paddingHorizontal: 15,
          marginEnd: 24,
        }}>
        <TextInput
          placeholder="Add a comment"
          multiline={true}
          style={styles.text}
          onChangeText={onChangeText}
          value={value}
        />
        <TouchableOpacity onPress={onPress}>
          <View
            style={{
              backgroundColor: '#00ADB5',
              paddingHorizontal: 16,
              paddingVertical: 4,
              borderRadius: 8,
              width: 80,
              height: 32,
              justifyContent: 'center',
              alignItems: 'center',
            }}>
            <Text style={{color: 'white', fontFamily: fonts.inter[400]}}>
              Reply
            </Text>
          </View>
        </TouchableOpacity>
      </View>
    </View>
  );
};

export default WriteComment;

const styles = StyleSheet.create({
  container: {
    flex: 1,
    borderTopWidth: 1,
    borderTopColor: colors.gray1,
    backgroundColor: '#fff',
    position: 'absolute',
    bottom: 0,
    width: '100%',
    paddingHorizontal: 22,
    paddingVertical: 10,
    flexDirection: 'row',
    alignItems: 'center',
  },
  image: {
    width: 32,
    height: 32,
  },
  text: {
    flex: 1,
  },
});
