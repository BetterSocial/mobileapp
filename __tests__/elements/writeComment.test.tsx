import React from 'react';
import {shallow, configure} from 'enzyme';
import toJson from 'enzyme-to-json';
import renderer from 'react-test-renderer';
import Adapter from 'enzyme-adapter-react-16';
import WriteComment from '../../src/elements/PostDetail/WriteComment';

configure({adapter: new Adapter()});

describe('component WriteComment', () => {
  it('WriteComment snapshot', () => {
    const component = shallow(
      <WriteComment
        onPress={() => {}}
        value={'halo'}
        onChangeText={() => {}}
      />,
    );
    expect(toJson(component)).toMatchSnapshot();
  });

  it('WriteComment renders correctly', () => {
    const tree = renderer
      .create(
        <WriteComment
          onPress={() => {}}
          value={'halo'}
          onChangeText={() => {}}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
