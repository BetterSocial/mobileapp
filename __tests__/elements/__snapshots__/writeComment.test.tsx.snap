// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component WriteComment WriteComment renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "borderTopColor": "#C4C4C4",
      "borderTopWidth": 1,
      "bottom": 0,
      "flex": 1,
      "flexDirection": "row",
      "paddingHorizontal": 22,
      "paddingVertical": 10,
      "position": "absolute",
      "width": "100%",
    }
  }
>
  <Image
    source={
      Object {
        "testUri": "../../../src/assets/images/ProfileDefault.png",
      }
    }
    style={
      Object {
        "height": 32,
        "width": 32,
      }
    }
  />
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#F5F5F5",
        "borderRadius": 5,
        "flexDirection": "row",
        "marginEnd": 24,
        "marginLeft": 10,
        "paddingHorizontal": 15,
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      multiline={true}
      onChangeText={[Function]}
      placeholder="Add a comment"
      rejectResponderTermination={true}
      style={
        Object {
          "flex": 1,
        }
      }
      underlineColorAndroid="transparent"
      value="halo"
    />
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#00ADB5",
            "borderRadius": 8,
            "height": 32,
            "justifyContent": "center",
            "paddingHorizontal": 16,
            "paddingVertical": 4,
            "width": 80,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "white",
              "fontFamily": "Inter-Regular",
            }
          }
        >
          Reply
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`component WriteComment WriteComment snapshot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "borderTopColor": "#C4C4C4",
      "borderTopWidth": 1,
      "bottom": 0,
      "flex": 1,
      "flexDirection": "row",
      "paddingHorizontal": 22,
      "paddingVertical": 10,
      "position": "absolute",
      "width": "100%",
    }
  }
>
  <Image
    source={
      Object {
        "testUri": "../../../src/assets/images/ProfileDefault.png",
      }
    }
    style={
      Object {
        "height": 32,
        "width": 32,
      }
    }
  />
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#F5F5F5",
        "borderRadius": 5,
        "flexDirection": "row",
        "marginEnd": 24,
        "marginLeft": 10,
        "paddingHorizontal": 15,
      }
    }
  >
    <Component
      allowFontScaling={true}
      multiline={true}
      onChangeText={[Function]}
      placeholder="Add a comment"
      rejectResponderTermination={true}
      style={
        Object {
          "flex": 1,
        }
      }
      underlineColorAndroid="transparent"
      value="halo"
    />
    <ForwardRef
      onPress={[Function]}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#00ADB5",
            "borderRadius": 8,
            "height": 32,
            "justifyContent": "center",
            "paddingHorizontal": 16,
            "paddingVertical": 4,
            "width": 80,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "white",
              "fontFamily": "Inter-Regular",
            }
          }
        >
          Reply
        </Text>
      </View>
    </ForwardRef>
  </View>
</View>
`;
